DC editor

jhadvig opened the pull request:
There is the DC editor work I've been working on.
The editor consists from three parts:
- Deployment Strategy Configuration
- Image Configuration
- Environment Variables

#### Deployment Strategy Configuration:
- serves for editing basic strategy parameters for each strategy:
  - Recreate - Timeout / Update Period / Interval / Max Unavailability / Max Surge
  - Rolling -  Timeout
  - Custom - Image name / Command / environment variables
- set `LifecycleHooks` for Recreate and Rolling strategy:
  - each hook might be for either:
      - Run a specific command in a new pod
      - Tag image if the deployment succeeds

#### Image Configuration:
- serves for setting manual or automatic trigger for each container in the podTemplate
- set of pullSecrets

#### Environment Variables:
Environment Variables for deployment


@spadgett @jwforres PTAL

jhadvig commented:
`dist` will be added after review

spadgett commented on app/scripts/directives/lifeCycleHook.js:
```diff
@@ -0,0 +1,113 @@
+        var setImageOptions = function(imageData) {
+          var istag = {};
+          if (!_.isEmpty(imageData)) {
+            istag = {namespace: imageData.namespace || $scope.namespace, imageStream: imageData.name.split(':')[0], tagObject: {tag: imageData.name.split(':')[1]}};
```
Don't split twice the string twice. This is more readable with newlines.

```js
var imageNameParts;
if (!_.isEmpty(imageData)) {
  imageNameParts = imageData.name.split(':');
  istag = {
    namespace: imageData.namespace || $scope.namespace,
    imageStream: imageNameParts[0],
    tagObject: {
      tag: imageNameParts[1]
    }
  };
} else {
  istag = {
    namespace: $scope.namespace,
    imageStream: "",
    tagObject: {
      tag: ""
    }
  };
}
```

spadgett commented on app/scripts/controllers/edit/deploymentConfig.js:
```diff
@@ -0,0 +1,289 @@
+        if (containerData.hasDeploymentTrigger) {
+          updatedTriggers.push(assembleImageChangeTrigger(containerName, containerData.triggerData.istag, containerData.triggerData.data));
+        } else {
+          _.find($scope.updatedDeploymentConfig.spec.template.spec.containers, function(container) {return container.name === containerName}).image = containerData.image;
```
Ah I see. I'd split this into two lines with a local var. Too easy to miss that you're setting the image value on the result.

spadgett commented on app/scripts/controllers/edit/deploymentConfig.js:
```diff
@@ -0,0 +1,289 @@
+    };
+
+    var updateTriggers = function() {
+      var updatedTriggers = _.remove($scope.updatedDeploymentConfig.spec.triggers, function(trigger) {return trigger.type !== "ImageChange"});
```
But the code seems to be removing anything that's *not* an ICT?

jhadvig replied:
It's kinds weird since but in fact it's vice versa... not sure if they have mistake in their docs but this will return array of non-ICT triggers

jhadvig replied:
anyway.. I've updated went with the `_.filter` :+1:

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,380 @@
+                          </div>
+                          <div class="help-block" id="image-name-help">Specifies a image which can carry out a deployment.</div>
+
+                          <div class="form-group gutter-top">
```
Is `gutter-top` needed if the field above is inside a `form-group`?

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,380 @@
+                              key-validator="[a-zA-Z][a-zA-Z0-9_]*"
+                              key-validator-error-tooltip="A valid environment variable name is an alphanumeric (a-z and 0-9) string beginning with a letter that may contain underscores."
+                              add-row-link="Add environment variable"></key-value-editor>
+                            <div class="help-block">Specifies environment which will be given to the container for Image.</div>
```
I'd just remove this help-block.

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,380 @@
+                              <span class="input-group-addon">seconds</span>
+                            </span>
+                            <div class="help-block" ng-attr-id="strategyTimeout">
+                              How long should deployment configuration retry to run the pod before times out.
```
Maybe

"How long to wait for a pod to scale up before giving up."

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,380 @@
+                                      aria-describedby="updatePeriod">
+                                  <span class="input-group-addon">seconds</span>
+                                </span>
+                                <div class="help-block" ng-attr-id="updatePeriod">
```
`id="updatePeriod"`, you don't need `ng-attr-` here

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,380 @@
+                                  <span class="input-group-addon">seconds</span>
+                                </span>
+                                <div class="help-block" ng-attr-id="updatePeriod">
+                                  Time to wait between retrying to run individual pod.
```
Is this correct? That sounds like timeout.

jhadvig replied:
should be right based on our api https://github.com/openshift/origin/blob/master/pkg/deploy/api/v1/types.go#L141

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,380 @@
+                                    Must be a number.
+                                  </div>
+                                  <div ng-if="form.interval.$error.min" class="help-block">
+                                    Delay can't be negative.
```
Interval can't be negative.

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,380 @@
+                                    Must be a number.
+                                  </div>
+                                  <div ng-if="form.updatePeriod.$error.min" class="help-block">
+                                    Delay can't be negative.
```
Update period can't be negative.

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,380 @@
+                                Must be a number.
+                              </div>
+                              <div ng-if="form.strategyTimeout.$error.min" class="help-block">
+                                Delay can't be negative.
```
Timeout can't be negative.

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,380 @@
+                                    placeholder="25%"
+                                    name="maxUnavailable"
+                                    ng-model="strategyData[displayedParams].maxUnavailable"
+                                    ng-pattern="/^\d+[%]{0,1}$/"
```
Does this regex work?

`/^\d+%?$/`

jhadvig replied:
it does

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,380 @@
+                              </div>
+                              <div class="help-block" id="max-unavailable-help">
+                                Maximum number of pods that can be unavailable during the update. Can be set to either a percentage (e.g., 10%) or an absolute value (e.g., 2).
+                                <a href="{{'rolling_strategy' | helpLink}}" aria-hidden="true" target="_blank"><span class="learn-more-inline">Learn more&nbsp;<i class="fa fa-external-link"></i></span></a>
```
Remove aria-hidden from the a element. Add aria-hidden to the i element.

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,380 @@
+                              </div>
+                              <div class="help-block" ng-attr-id="maxSurge">
+                                Maximum number of pods that can be scheduled above the original number of pods. Can be set to either a percentage (e.g., 10%) or an absolute value (e.g., 2).
+                                <a href="{{'rolling_strategy' | helpLink}}" aria-hidden="true" target="_blank"><span class="learn-more-inline">Learn more&nbsp;<i class="fa fa-external-link"></i></span></a>
```
Remove aria-hidden from the a element. Add aria-hidden to the i element.

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,380 @@
+                              <div ng-if="form.maxUnavailable.$invalid && form.maxUnavailable.$touched" class="has-error">
+                                <span ng-if="form.maxUnavailable.$error.pattern && !form.maxUnavailable.$error.min" class="help-block">
+                                  Must be a whole number or percentage.
+                                </span>
```
Do you need another error message for `$error.min`?

jhadvig replied:
since the input is text, because of the `%` sign that can occur in the input field, we can not check for `min`

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,380 @@
+                          </div>
+                        </div>
+                        <div class="gutter-top">
+                          <div ng-if="!view.advancedStrategyOptions">To set additional parameters or edit lifecycle hooks, switch to <a href="" ng-click="view.advancedStrategyOptions = true">advanced strategy configuration options.</a></div>
```
```
To set additional parameters or edit lifecycle hooks, view <a href="" ...>advanced options</a>.
```

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,380 @@
+                        </div>
+                        <div class="gutter-top">
+                          <div ng-if="!view.advancedStrategyOptions">To set additional parameters or edit lifecycle hooks, switch to <a href="" ng-click="view.advancedStrategyOptions = true">advanced strategy configuration options.</a></div>
+                          <a ng-if="view.advancedStrategyOptions" href="" ng-click="view.advancedStrategyOptions = false">Hide advanced strategy configuration options.</a>
```
Hide advanced options.

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,380 @@
+
+
+                    <div class="section">
+                      <h3 class="with-divider">Image Configuration</h3>
```
Maybe just 

```
<h3 class="with-divider">Images</h3>
```

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,380 @@
+                            <h4>Container {{containerName}}</h4>
+                          </div>
+                          <div class="form-group" ng-if="containerData.hasDeploymentTrigger">
+                            <label class="required">Trigger image</label>
```
Image Change Trigger

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,380 @@
+                          <div class="form-group" ng-if="containerData.hasDeploymentTrigger">
+                            <label class="required">Trigger image</label>
+                            <istag-select model="containerData.triggerData.istag" select-disabled="disableInputs" include-shared-namespace="true"></istag-select>
+                            <span class="help-block">New deployment will be automatically triggers once specified image is updated.</span>
```
"New deployments will automatically run when the image stream tag is updated."

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,380 @@
+                                spellcheck="false"
+                                required>
+                            </div>
+                            <span class="help-block">New deployment can be triggered only manually.</span>
```
I'd probably just remove this help block.

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,380 @@
+                          <div class="checkbox form-group">
+                            <label>
+                              <input type="checkbox" ng-model="containerData.hasDeploymentTrigger">
+                              Automatically trigger new deployments when image is updated.
```
"Automatically start new deployments when the image stream tag is updated."

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,380 @@
+                        </div>
+
+                        <div class="gutter-top">
+                          <div ng-if="!view.advancedImageOptions">To set secrets for pulling your images from private image registries, switch to <a href="" ng-click="view.advancedImageOptions = true">advanced image configuration options.</a></div>
```
```
...view <a href="" ...>advanced options</a>.
```

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,380 @@
+
+                        <div class="gutter-top">
+                          <div ng-if="!view.advancedImageOptions">To set secrets for pulling your images from private image registries, switch to <a href="" ng-click="view.advancedImageOptions = true">advanced image configuration options.</a></div>
+                          <a ng-if="view.advancedImageOptions" href="" ng-click="view.advancedImageOptions = false">Hide advanced image configuration options.</a>
```
Hide advanced options.

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,380 @@
+                        </a>
+                      </span></h3>
+
+                      <div ng-repeat="(containerName, containerData) in containersDataMap" class="">
```
`class=""`?

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,380 @@
+
+                    <div class="section">
+                      <h3 class="with-divider">Environment Variables<span class="help action-inline">
+                        <a href="">
```
I'd just remove the surround `a` element and keep the `i` as the tooltip.

jhadvig commented on app/scripts/directives/lifeCycleHook.js:
```diff
@@ -0,0 +1,113 @@
+          // Assamble image name when tag changes
+          _.set($scope.hookParams, ['tagImages', '0', 'to', 'kind'], 'ImageStreamTag');
+          _.set($scope.hookParams, ['tagImages', '0', 'to', 'namespace'], $scope.istagHook.namespace);
+          _.set($scope.hookParams, ['tagImages', '0', 'to', 'name'], $scope.istagHook.imageStream + ':' + $scope.istagHook.tagObject.tag);
```
I thought that we agreed that we will only edit the first IST (based on out blueJeans debate), since  I was told that most of the cases contain single IST in `tagImages`. But if you would like to I can add the logic that user can add/remove IST. Just let me know :)

spadgett replied:
OK, but let's make sure we don't drop any others in the array. It might be better to show the values and make it read-only if there is more than one.

jhadvig commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,380 @@
+                        {{strategyParamsName}}
+                        <div class="form-group strategy-name">
+                          <label class="picker-label">Strategy Type</label>
+                          <ui-select ng-model="strategyData.type" search-enabled="false" ng-change="strategyChange(strategyData.type)">
```
yep it can

spadgett commented:
Then you should have only one `ng-if` on the parent div if we only handle
one error.

Instead of

```html
<div ng-if="..." class="has-error">
  <span ng-if="..." class="help-block">...</span>
</div>
```

use

```html
<div ng-if="(all conditions here)" class="has-error">
  <span class="help-block">...</span>
</div>
```

On Wed, Oct 12, 2016 at 10:25 AM, Jakub Hadvig <notifications@github.com>
wrote:

> *@jhadvig* commented on this pull request.
> ------------------------------
>
> In app/views/edit/deployment-config.html
> <https://github.com/openshift/origin-web-console/pull/656>:
>
> > +                                    name="maxUnavailable"
> +                                    ng-model="strategyData[displayedParams].maxUnavailable"
> +                                    ng-pattern="/^\d+[%]{0,1}$/"
> +                                    min="0"
> +                                    select-on-focus
> +                                    class="form-control"
> +                                    aria-describedby="max-unavailable-help">
> +                              </div>
> +                              <div class="help-block" id="max-unavailable-help">
> +                                Maximum number of pods that can be unavailable during the update. Can be set to either a percentage (e.g., 10%) or an absolute value (e.g., 2).
> +                                <a href="{{'rolling_strategy' | helpLink}}" aria-hidden="true" target="_blank"><span class="learn-more-inline">Learn more&nbsp;<i class="fa fa-external-link"></i></span></a>
> +                              </div>
> +                              <div ng-if="form.maxUnavailable.$invalid && form.maxUnavailable.$touched" class="has-error">
> +                                <span ng-if="form.maxUnavailable.$error.pattern && !form.maxUnavailable.$error.min" class="help-block">
> +                                  Must be a whole number or percentage.
> +                                </span>
>
> since the input is text, because of the % sign that can occur in the
> input field, we can not check for min
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin-web-console/pull/656>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/ABHPm_k--48pqkpO9biTZhz79XEKu7JGks5qzO29gaJpZM4KS3oN>
> .
>

jhadvig commented:
@spadgett I've update the PR. PTAL

spadgett commented:
Changing strategy type values several times, I hit this error:

```
Error: [ui.select:choices] Expected multiple .ui-select-choices-row but got '0'.
    at select.js:128
    at _ensureHighlightVisible (select.js:978)
    at select.js:420
    at angular.js:16414
    at completeOutstandingRequest (angular.js:4940)
    at angular.js:5328
```

spadgett commented:
`x` button is not vertically centered with the select here:

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/19319805/9535d416-907c-11e6-954f-a868fcf3285f.png)

spadgett commented:
We don't need this many learn more links, particularly since they point to the same topic. I'd keep only the first.

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/19320079/d583a394-907d-11e6-81e1-ad90c9c7ab00.png)

spadgett commented:
The spacing seems off here. The help text should not be closer to the link below than the input it's describing.

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/19320897/2c0d1404-9081-11e6-9e0e-0a428bc6a57c.png)

spadgett commented:
We should use `&nbsp;` between "Learn more" and the icon instead of a normal space to keep the icon from wrapping on its own.

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/19322037/73ec2d10-9085-11e6-8e36-9ffa33177b5b.png)

spadgett commented:
These need to be checkboxes instead of radio buttons because you can have both set.

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/19322069/a1494b58-9085-11e6-8023-d244263df6ff.png)

"Run a specific command" should be checked by default for new hooks and validate that at least one of the checkboxes is selected.

spadgett commented:
We should warn the user if switching strategy will remove one of their hooks. For instance, switching from recreate to rolling will remove the mid hook. Or switching to custom will remove all hooks. But only if they have hooks defined.

spadgett commented:
These actions should be "Add another secret | Create new secret" to make it more obvious how they are different.

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/19322495/51189952-9087-11e6-8d07-b7e3a352395a.png)

jhadvig commented:
@spadgett I've added commit(will squash afterwards) which is addressing most of yours comments.

> We should use &nbsp; between "Learn more" and the icon instead of a normal space to keep the icon from wrapping on its own.

I've already have the `&nbsp;` in the place for all the `Learn more` links. Not sure what could cause the issue :-/

> We should warn the user if switching strategy will remove one of their hooks. For instance, switching from recreate to rolling will remove the mid hook. Or switching to custom will remove all hooks. But only if they have hooks defined.

I've updated the dialog msg which is now saying:
"The timeout parameter and any pre or post lifecycle hooks will be copied from recreate strategy to rolling strategy. After saving the changes, recreate strategy parameters will be removed ."
This only applies for the Rolling <-> Recreate switch. We are not removing any params when switching to Custom strategy, which was part of the debate we had with @jwforres

spadgett commented:
But we want to warn them if a hook that exists will not longer run. That still applies to the custom strategy, which doesn't support the hooks.

spadgett commented on app/scripts/controllers/edit/deploymentConfig.js:
```diff
@@ -0,0 +1,297 @@
+
+    $scope.strategyChanged = function() {
+      var pickedStrategyParams = getParamsPropertyName($scope.strategyData.type);
+      if (isRollingRecreateSwitch()) {
```
I'd pull everything inside this `if` into a helper function, maybe `promptToMoveParams`. It'd be easier to follow the if/else logic in strategyChanged.

spadgett commented on app/styles/_core.less:
```diff
@@ -1132,7 +1159,7 @@ a.disabled-link {
   }
   .checkbox {
-    margin: 10px;
+    margin: 10px 0px;
```
Use `0` instead of `0px`. 0 values don't need a unit.

spadgett commented on app/styles/_core.less:
```diff
@@ -503,18 +503,34 @@ label.checkbox {
   }
-  .directory {
-    width:90%;
+  .secret-name(@width: 90%) {
```
I think I'd rather just have

```css
.secret-name {
  width: 90%;
  display: inline-block;
}
```

and later

```css
.basic-secrets {
  .secret-name {
    width: 95%;
  }
}
```

I don't think the mixin is necessary?

spadgett commented on app/views/directives/lifecycle-hook.html:
```diff
@@ -0,0 +1,151 @@
+        <div class="form-group">
+          <label class="required">Tag As</label>
+          <istag-select model="istagHook" allow-custom-tag="true" select-disabled="view.isDisabled"></istag-select>
+        </div> -->
```
Remove commented out code?

spadgett commented on app/views/directives/lifecycle-hook.html:
```diff
@@ -0,0 +1,151 @@
+            </div>
+            <p ng-if="hookParams.tagImages.length > 1">
+              <span class="pficon pficon-info" aria-hidden="true"></span>
+              More then one image tags are defined in this deployment configuration. Only the first one is editable, below is displayed the rest.
```
It's weird that the first one only is editable if there is more than one. Either all image tags should be read-only (when there are multiple) or we should just try to support editing all of them.

spadgett commented on app/styles/_core.less:
```diff
@@ -503,18 +503,34 @@ label.checkbox {
+  .remove-secret {
+    display: inline-block;
+    margin-left: 7px;
+    vertical-align: 0px;
```
You and I are seeing very different behavior for `vertical-align` with this icon :(

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,373 @@
+                            </div>
+                          </div>
+                          <div class="gutter-top">
+                            <div ng-if="!view.advancedStrategyOptions">To set additional parameters or edit lifecycle hooks, view <a href="" ng-click="view.advancedStrategyOptions = true">advanced strategy options</a></div>
```
Here I'd have the period after the link since the entire sentence is not an action link, but I'd leave the period off for the hide link.

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,373 @@
+
+                          </div>
+
+                          <div ng-if="view.advancedImageOptions">
```
Since there's only one field we're hiding and showing, I wonder if it's even worth having the advanced link. (I know that's different than what we told you.)

@jwforres opinion?

jhadvig replied:
imho I would keep it hidden. The idea of having "basic/advanced view" is to compress the default view to bare minimum, for both BC and DC editor

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,373 @@
+                      <div class="section">
+                        <h3 class="with-divider">
+                          Environment Variables
+                          <span class="help action-inline"></span>
```
Remove this empty `span`

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,373 @@
+                          <span class="help action-inline"></span>
+                        </h3>
+
+                        <div ng-repeat="(containerName, containerData) in containerConfigByName">
```
nit: I'd probably call this `containerConfig` to match the map name.

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,373 @@
+                          type="submit"
+                          class="btn btn-primary btn-lg"
+                            ng-disabled="form.$invalid || form.$pristine || disableInputs">
+                            Save
```
indentation

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,373 @@
+                                  aria-describedby="image-name-help"
+                                  required>
+                              </div>
+                              <div class="help-block" id="image-name-help">Specifies a image which can carry out a deployment.</div>
```
"An image that can carry out the deployment."

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,373 @@
+                            <div>
+                              <span ng-switch="strategyData.type">
+                                <span class="help-block" ng-switch-when="Recreate">
+                                  Recreate strategy has basic rollout behavior and supports lifecycle hooks for injecting code into the deployment process.
```
I'd start each of these strategy descriptiong with "The".

"The recreate strategy has basic..."

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,373 @@
+                                </div>
+                                <div ng-if="form.maxUnavailable.$invalid && form.maxUnavailable.$touched && form.maxUnavailable.$error.pattern" class="has-error">
+                                  <span class="help-block">
+                                    Must be a non negative whole number or percentage.
```
"non-negative"

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,373 @@
+                                </div>
+                                <div ng-if="form.maxSurge.$invalid && form.maxSurge.$touched && form.maxSurge.$error.pattern" class="has-error">
+                                  <span class="help-block">
+                                    Must be a non negative whole number or percentage.
```
"non-negative"

spadgett commented:
@jhadvig Can you add the `pointer-events` line and comment to this rule in _select.less? It fixes the ui-select error for me.

```less
  .ui-select-toggle > .caret {
    font-size: (@font-size-base - 1);
    font-style: normal;
    right: 6px;
    top: 10px;
    // Workaround for https://github.com/angular-ui/ui-select/issues/224
    pointer-events: none;
  }
```

spadgett commented on app/scripts/controllers/edit/deploymentConfig.js:
```diff
@@ -157,39 +157,40 @@ angular.module('openshiftConsole')
     };
 
+    var promptToMoveParams = function(pickedStrategyParams) {
+      if (!_.has($scope.strategyData, pickedStrategyParams)) {
```
I prefer an early return rather than wrapping everything in an `if`. It's just a little easier to follow.

```js
if (_.has($scope.strategyData, pickedStrategyParams)) {
  return;
}

// Show modal...
```

spadgett commented on app/scripts/controllers/edit/deploymentConfig.js:
```diff
@@ -157,39 +157,40 @@ angular.module('openshiftConsole')
-            $scope.strategyData[pickedStrategyParams] = {};
-          });
-        }
+        promptToMoveParams(pickedStrategyParams);
```
Thanks, I think this is a lot easier to follow.

spadgett commented on app/styles/_core.less:
```diff
@@ -503,34 +503,40 @@ label.checkbox {
+    .remove-secret {
+      display: inline-block;
+      margin-left: 7px;
+      vertical-align: 8px;
```
👍 

glad we figured this out

spadgett commented on app/styles/_select.less:
```diff
@@ -45,6 +45,10 @@
+    // Workaround for https://github.com/angular-ui/ui-select/issues/224
+    // A runtime error occurs when you click exactly on the down caret for some
+    // selects, such as the strategy type select in the deployment config editor.
+    pointer-events: none;
```
👍

spadgett commented on app/views/directives/lifecycle-hook.html:
```diff
@@ -115,13 +96,14 @@
+            <p class="read-only-info" ng-if="$first">
               <span class="pficon pficon-info" aria-hidden="true"></span>
-              More then one image tags are defined in this deployment configuration. Only the first one is editable, below is displayed the rest.
+              More then one image tags are defined in this deployment configuration. To edit them, use deployment config YAML editor.
```
"More than one image tag is defined. To change image tags, use the YAML editor."

spadgett commented on app/views/directives/lifecycle-hook.html:
```diff
@@ -115,13 +96,14 @@
-          </div>
-          <div class="gutter-bottom read-only-tag-image" ng-if="!$first">
-            <span class="container-name">{{tagImage.containerName}}</span><i class="fa fa-long-arrow-right"></i><span class="image-tag">{{tagImage.to.namespace || namespace}}/{{tagImage.to.name}}</span>
+            <span class="container-name">{{tagImage.containerName}}</span>&rarr;<span class="image-tag">{{tagImage.to.namespace || namespace}}/{{tagImage.to.name}}</span>
```
I'd remove the `container-name` and `image-tag` styles and just use this:

```html
{{tagImage.containerName}}&nbsp;&rarr;&nbsp;{{tagImage.to.namespace || namespace}}/{{tagImage.to.name}}
```

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,366 @@
+                                  spellcheck="false"
+                                  required>
+                              </div>
+                              <span class="help-block">New deployment can be triggered only manually.</span>
```
This isn't really true since there could be config change triggers as well. I'd just remove the help here.

spadgett commented on app/views/edit/deployment-config.html:
```diff
@@ -0,0 +1,366 @@
+                                <input type="checkbox" ng-model="containerConfig.hasDeploymentTrigger">
+                                Automatically start new deployments when the image stream tag is updated.
+                              </label>
+                            </div>
```
Forgot to ask you do this, but we should move this above the ICT and Image Name fields.

spadgett commented:
Just a few small requests. Will need to squash commits and add dist before we merge.

We should have a config change trigger toggle to the editor, but that can be a follow on.

spadgett commented:
Needs more top margin above the second container name:

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/19387804/93a3c850-91ea-11e6-82f4-d34b7b5ac72d.png)

spadgett commented:
When I clicked the "Automatically start new deployments when the image stream tag is updated." the dropdown prefilled an image stream that was invalid.

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/19387888/f1b3ec90-91ea-11e6-9416-2421c7bbb14a.png)

jhadvig commented:
@spadgett regarding your last comment. The problem is that we can not determine the Tag of the IST for the ICT from the imageSpec, in case switching from manual to automatic trigger, since the API doesn't support it yet. So I've skipped it and fill the istag directive with the namespace and imageStream, which we can get from the imageSpec. 
Other solution would be to only prefill the namespace, but that could be confusing cause the imageStream is defined.

spadgett commented:
>
> The problem is that we can not determine the Tag of the IST for the ICT
> from the imageSpec, in case switching from manual to automatic trigger,
> since the API doesn't support it yet.


It's not the tag. The image stream exists in a different namespace, so what
you prefilled is wrong.

jhadvig commented:
@spadgett I've updated the PR based on your comments + added `dist` + squashed
PTAL

spadgett commented:
[test]

spadgett commented:
[merge]
